mixer/pkg/runtime2/runtime.go

```diff
@@ -0,0 +1,195 @@
+// instances, creates the dispatch machinery and handles incoming requests.
+type Runtime struct {
+	identityAttribute string
+
+	defaultConfigNamespace string
+
+	ephemeral *config.Ephemeral
+
+	snapshot *config.Snapshot
+
+	handlers *handler.Table
+
+	dispatcher *dispatcher.Dispatcher
+
+	env adapter.Env
+
+	store store.Store
+
+	handlerPool *pool.GoroutinePool
+
+	stateLock            sync.Mutex
+	shutdown             chan struct{}
+	waitQuiesceListening sync.WaitGroup
+}
+
+// New returns a new instance of Runtime.
+func New(
+	s store.Store,
+	templates map[string]*template.Info,
+	adapters map[string]*adapter.Info,
+	identityAttribute string,
+	defaultConfigNamespace string,
+	executorPool *pool.GoroutinePool,
+	handlerPool *pool.GoroutinePool,
+	enableTracing bool) *Runtime {
+
+	runtime := &Runtime{
+		identityAttribute:      identityAttribute,
+		defaultConfigNamespace: defaultConfigNamespace,
+		ephemeral:              config.NewEphemeral(templates, adapters),
+		snapshot:               config.Empty(),
+		handlers:               handler.Empty(),
+		dispatcher:             dispatcher.New(identityAttribute, executorPool, enableTracing),
+		handlerPool:            handlerPool,
+
+		store: s,
+	}
+
+	// Make sure we have a stable state.
+	runtime.processNewConfig()
+
+	return runtime
+}
+
+// Dispatcher returns the runtime.Dispatcher that is implemented by this runtime package.
+func (c *Runtime) Dispatcher() runtime.Dispatcher {
+	return c.dispatcher
+}
+
+// StartListening directs Runtime to start listening to configuration changes. As config changes, runtime processes
+// the confguration and creates a dispatcher.
+func (c *Runtime) StartListening() error {
+	c.stateLock.Lock()
+	defer c.stateLock.Unlock()
+
+	if c.shutdown != nil {
+		return errors.New("already listening")
+	}
+
+	kinds := config.KindMap(c.snapshot.Adapters, c.snapshot.Templates)
+
+	data, watchChan, err := startWatch(c.store, kinds)
+	if err != nil {
+		return err
+	}
+
+	c.ephemeral.SetState(data)
+	c.processNewConfig()
+
+	c.shutdown = make(chan struct{})
+	c.waitQuiesceListening.Add(1)
+	go func() {
+		watchChanges(watchChan, c.shutdown, c.onConfigChange)
+		c.waitQuiesceListening.Done()
+	}()
+
+	return nil
+}
+
+// StopListening directs Runtime to stop listening to configuration changes. It will not unload the current
+// configuration, or close the existing adapters.
+func (c *Runtime) StopListening() {
+	c.stateLock.Lock()
+	defer c.stateLock.Unlock()
+
+	if c.shutdown != nil {
+		c.shutdown <- struct{}{}
+		c.shutdown = nil
+		c.waitQuiesceListening.Wait()
+	}
+}
+
+func (c *Runtime) onConfigChange(events []*store.Event) {
+	for _, e := range events {
+		c.ephemeral.ApplyEvent(e)
+	}
+	c.processNewConfig()
+}
+
+func (c *Runtime) processNewConfig() {
+	newSnapshot := c.ephemeral.BuildSnapshot()
+
+	oldHandlers := c.handlers
+
+	newHandlers := handler.NewTable(oldHandlers, newSnapshot, c.handlerPool)
+
+	builder := compiled.NewBuilder(newSnapshot.Attributes)
```

Comment:
is `builder` equivalent to `compiler`. Should we rename ?

Reply:
No, builder and compiler are different. builder is a CompiledExpressionBuilder.